grpc/read_job.cpp

```diff
@@ -0,0 +1,159 @@
+#include "read_job.hpp"
+
+#include "library/request_queue.h"
+#include "serialization.hpp"
+
+namespace ioremap { namespace grpc { namespace {
+
+void deserialize(const read_job_t::rpc_request_t &rpc_request, read_job_t::request_t &request) {
+	auto fb_request = rpc_request.GetRoot();
+
+	deserialize_cmd(fb_request->cmd(), request.cmd);
+	request.cmd.cmd = dnet_commands::DNET_CMD_READ_NEW;
+	request.ioflags = fb_request->ioflags();
+	request.read_flags = fb_request->read_flags();
+	request.data_offset = fb_request->data_offset();
+	request.data_size = fb_request->data_size();
+}
+
+flatbuffers::Offset<fb_grpc_dnet::ReadResponseHeader> serialize_header(flatbuffers::grpc::MessageBuilder &builder,
+	const read_job_t::response_t &response) {
+
+	auto json_timestamp = to_rpc_time(response.json_timestamp);
+	auto data_timestamp = to_rpc_time(response.data_timestamp);
+
+	return fb_grpc_dnet::CreateReadResponseHeader(
+		builder,
+		serialize_cmd(builder, response.cmd),
+		response.record_flags,
+		response.user_flags,
+		&json_timestamp,
+		response.json_size,
+		response.json_capacity,
+		response.read_json_size,
+		&data_timestamp,
+		response.data_size,
+		response.read_data_offset,
+		response.read_data_size
+	);
+}
+
+flatbuffers::Offset<flatbuffers::Vector<uint8_t>> put_data_part(flatbuffers::grpc::MessageBuilder &builder,
+	const elliptics::data_pointer &data, size_t &offset) {
+
+	if (offset == data.size() || builder.GetSize() >= GRPC_MAX_MESSAGE_SIZE) {
+		return 0;
+	}
+
+	size_t max_part_size = GRPC_MAX_MESSAGE_SIZE - builder.GetSize();
+	size_t size_left = data.size() - offset;
+	size_t part_size = std::min(size_left, max_part_size);
+
+	auto part_ptr = static_cast<const uint8_t *>(data.skip(offset).data());
+	auto res = builder.CreateVector(part_ptr, part_size);
+	offset += part_size;
+	return res;
+}
+
+read_job_t::rpc_response_t serialize_part(const read_job_t::response_t &response, bool first,
+	std::size_t &json_offset, std::size_t &data_offset) {
+
+	flatbuffers::grpc::MessageBuilder builder;
+
+	auto header = first
+	              ? serialize_header(builder, response)
+	              : flatbuffers::Offset<fb_grpc_dnet::ReadResponseHeader>();
+
+	builder.Finish(fb_grpc_dnet::CreateReadResponse(
+		builder,
+		header,
+		put_data_part(builder, response.json, json_offset),
+		put_data_part(builder, response.data.in_memory, data_offset)
+	));
+	return builder.ReleaseMessage<fb_grpc_dnet::ReadResponse>();
+}
+
+} // namespace
+
+class read_job_t::responder_t {
```

Comment:
I predict that for all the jobs `responder_t` will be exactly so trivial as he see here. I think it's excessive sugar with no future. Better to inherit all jobs from common responder. But it's only my opinion.

Reply:
There is not `n2_call` in repo. It's temporary decision.